src/app/tracing/services/highlighting.service.ts

```diff
@@ -177,39 +177,86 @@ export class HighlightingService {
+        );
+        const deliveryRulesToDisplay = this.enabledDelHRules.filter(rule =>
+            rule.showInLegend && (activeHighlightings.deliveries[rule.id])
         );
 
-        return {
-            stations: this.enabledStatHRules.filter(rule =>
-                rule.showInLegend &&
-                (activeHighlightings.stations[rule.id] || ruleIdToIsCommonLinkRuleMap[rule.id])
-            ).map(rule =>
-                ({ label: rule.name, color: this.mapToColor(rule.color), shape: rule.shape })
-            ),
-            deliveries: this.enabledDelHRules.filter(rule =>
-                rule.showInLegend && (activeHighlightings.deliveries[rule.id])
-            ).map(rule =>
-                ({ label: rule.name, color: this.mapToColor(rule.color), linePattern: rule.linePattern })
+        return this.statHighlightingRules
+            .map(rule => this.stationRuleToDisplayEntry(rule))
+            .concat(this.delHighlightingRules
+                .map(rule => this.deliveryRuleToDisplayEntry(rule))
             )
-        };
+            .reduce(
+                (array: LegendDisplayEntry[], current) => {
+                    const index = array.findIndex((entry) => entry.name === current.name);
+                    if (index >= 0) {
+                        //Should use legend.with(index,{newObj}) once we update to a newer TS version.
+                        const newArray = array;
+                        newArray[index] = {
+                            name: current.name,
+                            stationColor: array[index]?.stationColor ?? current?.stationColor,
+                            deliveryColor: array[index]?.deliveryColor ?? current?.deliveryColor,
+                            shape: array[index]?.shape ?? current?.shape
+                        };
+                        return newArray;
+                    } else {
+                        return [...array,
+                            {
+                                name: current.name,
+                                stationColor: array[index]?.stationColor ?? current?.stationColor,
+                                deliveryColor: array[index]?.deliveryColor ?? current?.deliveryColor,
+                                shape: array[index]?.shape ?? current?.shape
+                            }
+                        ];
+                    }
+                },
+                []
+            )
+            .map(rule => stationRulesToDisplay.some(enabledRule => enabledRule.name === rule.name)
+                ? rule
+                : { ...rule, stationColor: undefined, shape: undefined }
+
+            )
+            .map(rule => deliveryRulesToDisplay.some(enabledRule => enabledRule.name === rule.name)
+                ? rule
+                : { ...rule, deliveryColor: undefined }
+            )
+            .filter(rule => rule.deliveryColor || rule.shape || rule.stationColor);
```

Comment:
`rule.shape` is a enum type or undefined. This test assumes that enum fields are of type text. Otherwise `rule.shape`could be 0.
Please use `rule.shape !== undefined` instead